### ISSUE TITLE
[RFC] PEP 526: class member annotations in 3.6

### ISSUE THREAD

**Issue by hynek** [python-attrs/attrs#214]
I just played around and got:

```pycon
>>> import attr

>>> @attr.s
... class C:
...     x: int
...     y: int = attr.ib(default=42)

>>> C(1)
C(x=1, y=42)

>>> C("1", 23)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<attrs generated init 057634913447d4e148358419a9dd355e479f9105>", line 5, in __init__
    __attr_validator_x(self, __attr_x, self.x)
  File "/Users/hynek/Projects/attrs/src/attr/validators.py", line 33, in __call__
    attr, self.type, value,
TypeError: ("'x' must be <class 'int'> (got '1' that is a <class 'str'>).", Attribute(name='x', default=NOTHING, validator=<instance_of validator for type <class 'int'>>, repr=True, cmp=True, hash=None, init=True, convert=None, metadata=mappingproxy({})), <class 'int'>, '1')
("'x' must be <class 'int'> (got '1' that is a <class 'str'>).", Attribute(name='x', default=NOTHING, validator=<instance_of validator for type <class 'int'>>, repr=True, cmp=True, hash=None, init=True, convert=None, metadata=mappingproxy({})), <class 'int'>, '1')
```

working within five minutes (biggest “challenge” was to ignore the counter on _CountingAttr).

I think this is really neat because it saves us from fiddling with the class body and makes much more sense conceptually.

***

Now, obviously we can’t just start harvesting class annotations, we need an option or even a new decorator.

I also see this as a chance to get rid of the slots/non-slots dichotomy that fell a few times on our feet and *always* go the slots way (= create new class).

So I hereby ask for:

- feedback/opinions
- suggestions for decorator/option name
- other warts we could fix while doing this

**Comment by glyph** [comment]
Over on https://github.com/ericvsmith/dataclasses/issues/19 Guido made a very good point; if we want the type checker to work, the type field needs to have nothing but a type in it.  So: `x: int`, but `y = attr.ib()`; possibly `y: int = attr.ib()`.

**Comment by hynek** [comment]
(thanks, I’ve updated the comment so people don’t see the wrong version first thing)

**Comment by glyph** [comment]
Cool! I take it you got that working as well, without issue :)

**Comment by jmcs** [comment]
``y: int = attr.ib()`` doesn't work well with pylint (pylint will use the type of `ib` not `int`). For the common use case of setting a default I would like something like:

```python
>>> import attr

>>> @attr.s
... class C:
...     x: int
...     y: int = 42

>>> C(1)
C(x=1, y=42)
```

**Comment by glyph** [comment]
> y: int = attr.ib() doesn't work well with pylint (pylint will use the type of ib not int).

IMHO the solution here is to fix pylint; it's just doing the wrong thing.  This idea of defaults = cvars seems like a good idea until (as the participants in the dataclasses repo did) you realize that you will almost immediately want

> > `x: list = []`

and now nothing makes sense any more.  If you do

> > `x: list = new(list)`

it's still the wrong type.

**Comment by hynek** [comment]
FTR, I consider #215 a prerequisite for working on this.  In the end, this is just syntactic sugar for it.

**Comment by hynek** [comment]
PEP 526 support has been merged in #239!
